### PR TITLE
Support `AZURE_AUTH_LOCATION`

### PR DESCRIPTION
**Pull request by ellismg**
This change adds support to EnvironmentCredential so that it can
consider `AZURE_AUTH_LOCATION` as a location to get credentials. This
file can be generated with the Azure CLI.

The invasive part of the change here is refactoring things such that
instead of constructing the inner credential object eagerly when
`EnvironmentCredential` is constructed we build it on the first call to
GetToken or GetTokenAsync. This is done so that we do not have do
preform IO when constructing the credential object.

Contributes to #9309

### CODE REVIEW

**Comment by ellismg** [comment]
I'm not 100% sure on the design here, @schaabs, but I think this is good enough to land. I could image creating an actual SdkAuthCredential type that would be constructed with a file path, and EnvironmentCredential could create that. We could consider exposing that type as way to allow folks to use these files without setting environment variables.

The reason I didn't take that approach at first is that I believe that this new token would need to be an IExtendedTokenCredential and we would have to change EnvironmentCredential to handle calling the IExtendedTokenCredential when delegating to an underlying credential if they exist, in order to not end up double wrapping errors in some cases. It started to feel like a bunch of machinery and it wasn't clear it was worthwhile. I also know that you were already a little worried regarding the number of credentials types we had and throwing more public ones into the picture didn't seem so great.

Note that this does yet support a SDK Auth file that uses clientCertificate instead of clientSecret. When we end up supporting that (I assume the major blocker is going to be parsing the .pem file into an X509Certificate2) it will be easy to plug that same support in here.

**Review by jianghaolu — APPROVED** [review]
LGTM for auth files with a client secret

**Review by schaabs — APPROVED** [review]

**Comment by ellismg** [comment]
@schaabs I made the changes you and I discussed in person, could you please take another look?  Note that at this time I have SdkClientAuthCredential as public.  Do we need to do some surface area review?

File: sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs [hunk 1]
```diff
@@ -102,6 +102,13 @@ public partial class ManagedIdentityCredential : Azure.Core.TokenCredential
         public override Azure.Core.AccessToken GetToken(Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
         public override System.Threading.Tasks.ValueTask<Azure.Core.AccessToken> GetTokenAsync(Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) { throw null; }
     }
+    public partial class SdkAuthFileCredential : Azure.Core.TokenCredential
```
Comment (schaabs) [inline sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs]: Do you think the type name should be simplified to just `AuthFileCredential` prepending Sdk seems redundant as every class is in the SDK

Comment (ellismg) [reply sdk/identity/Azure.Identity/api/Azure.Identity.netstandard2.0.cs]: Works for me.

**Pull request merged**